### ISSUE TITLE
Mixerless telemetry memory leak

### ISSUE THREAD

**Issue by howardjohn** [istio/istio#17305]
I have a setup running `master` with mixerless telemetry. I see my ingressgateways memory usage grow over a ~1hr period to 500mb, at which point it is oom killed

Attached is a profile of memory usage. Seems `OwnedImpl::reserve` and `OwnedImpl::add` uses a lot
[profile.pb.gz](https://github.com/istio/istio/files/3640428/profile.pb.gz)

I will turn of mixerless for a bit and make sure that is the cause

**Comment by howardjohn** [comment]
I turned off mixerv2 and memory went back to a steady 100mb

**Comment by howardjohn** [comment]
Still impacted by 1.3.1

**Comment by mandarjog** [comment]
```
      flat  flat%   sum%        cum   cum%
    1248kB 25.39% 25.39%     1248kB 25.39%  Envoy::Buffer::OwnedImpl::reserve
     392kB  7.98% 33.37%      392kB  7.98%  Envoy::Buffer::OwnedImpl::add
  322.50kB  6.56% 39.93%   322.50kB  6.56%  Envoy::Http::HeaderString::setCopy
  275.94kB  5.61% 45.54%   275.94kB  5.61%  std::__1::list::emplace
  235.31kB  4.79% 50.33%   235.31kB  4.79%  Envoy::Http::HeaderMapImpl::HeaderList::insert
  220.82kB  4.49% 54.82%   220.82kB  4.49%  Envoy::Extensions::Common::Wasm::Wasm::allocMemory
  211.36kB  4.30% 59.12%   249.45kB  5.08%  std::__1::shared_ptr::make_shared
  169.07kB  3.44% 62.56%   199.57kB  4.06%  Envoy::Http::ConnectionManagerImpl::newStream
  137.32kB  2.79% 65.35%   374.76kB  7.62%  Envoy::Http::Http1::ConnectionImpl::onMessageBeginBase
  136.80kB  2.78% 68.14%  1369.07kB 27.85%  std::__1::__function::__func::operator()
```

**Comment by mandarjog** [comment]
![profile001](https://user-images.githubusercontent.com/800375/65808171-8150e200-e147-11e9-9074-e1614e886d35.png)

**Comment by mandarjog** [comment]
In another profile
```
File: envoy
Type: inuse_space
Showing nodes accounting for 15270.84kB, 92.56% of 16497.57kB total
Dropped 482 nodes (cum <= 82.49kB)
      flat  flat%   sum%        cum   cum%
 2137.90kB 12.96% 12.96%  2137.90kB 12.96%  Envoy::Extensions::Common::Wasm::Wasm::allocMemory
 1631.73kB  9.89% 22.85%  1631.97kB  9.89%  std::__1::vector::__push_back_slow_path
 1267.07kB  7.68% 30.53%  1428.73kB  8.66%  google::protobuf::Arena::CreateMaybeMessage
 1253.48kB  7.60% 38.13%  1253.48kB  7.60%  std::__1::basic_string::basic_string
 1172.50kB  7.11% 45.24%  1172.50kB  7.11%  std::__1::basic_string::append
  892.50kB  5.41% 50.64%  2324.08kB 14.09%  Envoy::Router::VirtualHostImpl::VirtualHostImpl
  784.95kB  4.76% 55.40%   784.95kB  4.76%  google::protobuf::Map::Init
     644kB  3.90% 59.31%      644kB  3.90%  Envoy::Buffer::OwnedImpl::reserve
  541.73kB  3.28% 62.59%   617.40kB  3.74%  std::__1::shared_ptr::make_shared
  434.84kB  2.64% 65.23%   635.13kB  3.85%  google::protobuf::Map::CreateValueTypeInternal
  361.16kB  2.19% 67.42%  1360.33kB  8.25%  std::__1::__hash_table::__emplace_unique_key_args
  346.50kB  2.10% 69.52%   346.50kB  2.10%  Envoy::Router::RateLimitPolicyImpl::RateLimitPolicyImpl
  338.48kB  2.05% 71.57%   338.48kB  2.05%  google::protobuf::internal::ArenaStringPtr::CreateInstanceNoArena
  304.50kB  1.85% 73.41%  2743.75kB 16.63%  Envoy::Router::RouteMatcher::RouteMatcher
  302.42kB  1.83% 75.25%   302.42kB  1.83%  hist_alloc
```

**Comment by mandarjog** [comment]
```
Dropped 607 nodes (cum <= 1.04MB)
Showing top 10 nodes out of 192
      flat  flat%   sum%        cum   cum%
   29.17MB 14.03% 14.03%    29.17MB 14.03%  Envoy::Buffer::OwnedImpl::add
   21.49MB 10.34% 24.36%    21.49MB 10.34%  hist_alloc
   20.25MB  9.74% 34.10%    20.26MB  9.74%  Envoy::Buffer::OwnedImpl::reserve
   14.21MB  6.83% 40.93%    14.21MB  6.83%  Envoy::Stats::FakeSymbolTableImpl::populateList
   12.68MB  6.10% 47.03%    12.68MB  6.10%  std::__1::basic_string::basic_string
   10.33MB  4.97% 52.00%    10.33MB  4.97%  Envoy::Extensions::Common::Wasm::Wasm::allocMemory
    7.44MB  3.58% 55.58%     7.44MB  3.58%  Envoy::Stats::FakeSymbolTableImpl::encode
```

**Comment by howardjohn** [comment]
Confirmed its not just ingressgateway, that was just due to gateway having more traffic I think

**Comment by mandarjog** [comment]
Running experiments with ASAN to look for mem leak.
So far 3 cases, but *no* mem leak, process dies before reaching it I suppose.

```
==1153==ERROR: AddressSanitizer: heap-use-after-free on address 0x611000651840 at pc 0x0000135e075c bp 0x7f4edc3e57f0 sp 0x7f4edc3e57e8
READ of size 8 at 0x611000651840 thread T136
    #0 0x135e075b in Envoy::Extensions::Common::Wasm::Wasm::Wasm(Envoy::Extensions::Common::Wasm::Wasm const&, Envoy::Event::Dispatcher&) /proc/self/cwd/external/envoy/source/extensions/common/wasm/wasm.cc:2384:29
.
.
0x611000651840 is located 0 bytes inside of 224-byte region [0x611000651840,0x611000651920)
freed by thread T0 here:
    #0 0x10daf662 in __interceptor_free /build/llvm-toolchain-8-8.0.1~svn363027/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:124:3
    #1 0x1a11c247 in Envoy::Stats::ThreadLocalStoreImpl::ScopeImpl::~ScopeImpl() /proc/self/cwd/external/envoy/source/common/stats/thread_local_store.cc:263:47
    #2 0x129be095 in std::__1::default_delete<Envoy::Stats::Scope>::operator()(Envoy::Stats::Scope*) const /usr/lib/llvm-8/bin/../include/c++/v1/memory:2339:5
    #3 0x129bdde5 in std::__1::unique_ptr<Envoy::Stats::Scope, std::__1::default_delete<Envoy::Stats::Scope> >::reset(Envoy::Stats::Scope*) /usr/lib/llvm-8/bin/../include/c++/v1/memory:2652:7
    #4 0x129b0b42 in std::__1::unique_ptr<Envoy::Stats::Scope, std::__1::default_delete<Envoy::Stats::Scope> >::~unique_ptr() /usr/lib/llvm-8/bin/../include/c++/v1/memory:2606:19
    #5 0x19b1e860 in Envoy::Server::ListenerImpl::~ListenerImpl() /proc/self/cwd/external/envoy/source/server/listener_manager_impl.cc:330:1

.
.
previously allocated by thread T0 here:
    #0 0x10daf9e3 in __interceptor_malloc /build/llvm-toolchain-8-8.0.1~svn363027/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:146:3
    #1 0x1f4c1799 in operator new(unsigned long) libcxx/build/lib/./libcxx/src/new.cpp:72:17
    #2 0x1a115d21 in Envoy::Stats::ThreadLocalStoreImpl::createScope(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) /proc/self/cwd/external/envoy/source/common/stats/thread_local_store.cc:109:20
    #3 0x19b18583 in Envoy::Server::ListenerImpl::ListenerImpl(envoy::api::v2::Listener const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, Envoy::Server::ListenerManagerImpl&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, bool, bool, unsigned long, Envoy::ProtobufMessage::ValidationVisitor&) /proc/self/cwd/external/envoy/source/server/listener_manager_impl.cc:196:45
    #4 0x19b271f7 in Envoy::Server::ListenerManagerImpl::addOrUpdateListener(envoy::api::v2::Listener const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, bool) /proc/self/cwd/external/envoy/source/server/listener_manager_impl.cc:514:36
    #5 0x19dbf135 in Envoy::Server::LdsApiImpl::onConfigUpdate(google::protobuf::RepeatedPtrField<envoy::api::v2::Resource> const&, google::protobuf::RepeatedPtrField<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) /proc/self/cwd/external/envoy/source/server/lds_api.cc:59:29
.
.
SUMMARY: AddressSanitizer: heap-use-after-free /proc/self/cwd/external/envoy/source/extensions/common/wasm/wasm.cc:2384:29 in Envoy::Extensions::Common::Wasm::Wasm::Wasm(Envoy::Extensions::Common::Wasm::Wasm const&, Envoy::Event::Dispatcher&)

```


```
==23==ERROR: AddressSanitizer: heap-use-after-free on address 0x604003acceb8 at pc 0x000011765578 bp 0x7f3e6b76a8b0 sp 0x7f3e6b76a8a8
READ of size 8 at 0x604003acceb8 thread T136
2019-10-01T14:24:18.591342Z     info    Envoy proxy is ready
    #0 0x11765577 in std::__1::unique_ptr<Envoy::ThreadLocal::Slot, std::__1::default_delete<Envoy::ThreadLocal::Slot> >::operator->() const /usr/lib/llvm-8/bin/../include/c++/v1/memory:2621:19
    #1 0x19fe9697 in Envoy::Router::RdsRouteConfigProviderImpl::onConfigUpdate()::$_2::operator()() const /proc/self/cwd/external/envoy/source/common/router/rds_impl.cc:198:38    #2 0x19fe9546 in _ZNSt3__18__invokeIRZN5Envoy6Router26RdsRouteConfigProviderImpl14onConfigUpdateEvE3$_2JEEEDTclclsr3std3__1E7forwardIT_Efp_Espclsr3std3__1E7forwardIT0_Efp0_EEEOS6_DpOS7_ /usr/lib/llvm-8/bin/../include/c++/v1/type_traits:4353:1
.
.
0x604003acceb8 is located 40 bytes inside of 48-byte region [0x604003acce90,0x604003accec0)
freed by thread T0 here:
    #0 0x10daf662 in __interceptor_free /build/llvm-toolchain-8-8.0.1~svn363027/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:124:3
    #1 0x19fce587 in Envoy::Router::RdsRouteConfigProviderImpl::~RdsRouteConfigProviderImpl() /proc/self/cwd/external/envoy/source/common/router/rds_impl.cc:186:59
.
.
previously allocated by thread T0 here:
    #0 0x10daf9e3 in __interceptor_malloc /build/llvm-toolchain-8-8.0.1~svn363027/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:146:3
    #1 0x1f4c1799 in operator new(unsigned long) libcxx/build/lib/./libcxx/src/new.cpp:72:17
    #2 0x19fc6ce0 in Envoy::Router::RouteConfigProviderUtil::create(envoy::config::filter::network::http_connection_manager::v2::HttpConnectionManager const&, Envoy::Server::Configuration::FactoryContext&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, Envoy::Router::RouteConfigProviderManager&) /proc/self/cwd/external/envoy/source/common/router/rds_impl.cc:33:42
.
SUMMARY: AddressSanitizer: heap-use-after-free /usr/lib/llvm-8/bin/../include/c++/v1/memory:2621:19 in std::__1::unique_ptr<Envoy::ThreadLocal::Slot, std::__1::default_delete<Envoy::ThreadLocal::Slot> >::operator->() const
```


```
==166==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x604004f0a638 at pc 0x000011765578 bp 0x7f18cef8c8b0 sp 0x7f18cef8c8a8
READ of size 8 at 0x604004f0a638 thread T136
2019-10-01T14:26:45.344770Z     info    Envoy proxy is ready
    #0 0x11765577 in std::__1::unique_ptr<Envoy::ThreadLocal::Slot, std::__1::default_delete<Envoy::ThreadLocal::Slot> >::operator->() const /usr/lib/llvm-8/bin/../include/c++/v1/memory:2621:19
    #1 0x19fe9697 in Envoy::Router::RdsRouteConfigProviderImpl::onConfigUpdate()::$_2::operator()() const /proc/self/cwd/external/envoy/source/common/router/rds_impl.cc:198:38
.
.
Address 0x604004f0a638 is a wild pointer.
SUMMARY: AddressSanitizer: heap-buffer-overflow /usr/lib/llvm-8/bin/../include/c++/v1/memory:2621:19 in std::__1::unique_ptr<Envoy::ThreadLocal::Slot, std::__1::default_delete<Envoy::ThreadLocal::Slot> >::operator->() const
```

**Comment by mandarjog** [comment]
[wasm_leak.zip](https://github.com/istio/istio/files/3677578/wasm_leak.zip)

**Comment by jplevyak** [comment]
@jmarantz see the ASAN Stats::Scope use-after-free above.

**Comment by jmarantz** [comment]
can you repro the use-after-free? If so I will take a look.

Better yet, can we code-inspect to make sure the scope is instantiated before the objects that reference it, in the relevant contexts?

**Comment by jplevyak** [comment]
@jmarantz so the problem is that reconfiguration results in ListenerImpl being destroyed which destroys Scope, but the way reconfiguration works, tls slots are Set asynchronously, and part of the data which is sent to the std::function via the Dispatcher to the other thread is a reference to the Scope.  If two reconfigurations race then the ListenerImpl can be destroyed before the tls Set has finished on all the workers.

Envoy could use the warming state for this, but AFAICT most extensions are not so I believe it is a more general problem.  std::shared_ptr is also a solution.

**Comment by jmarantz** [comment]
For shared_ptr. which Scope& should be managed in a shared_ptr? How will it be retained in between the destruction of the old and the creation of the new?

I think if we're going to capture a stat or 'this' in a std::function we should probably just defer deletion of the ListenerImpl until the std::function runs. WDYT?

**Comment by lambdai** [comment]
The magic is not attributed to warm up. 
ListenerImpl is destructed only if all the connections are drained. Search for `DrainingListener`
If wasm is allowed to access ListenerImpl after drained, more than scope could be broken.

My point is not design anything just to resolve scope

**Comment by rshriram** [comment]
Can we also isolate the issues as well? There are two filters at play here. the metadata exchange filter and the stats filter. If only metadata exchange is enabled, do we still see the memory leak?

**Comment by mandarjog** [comment]
The issue is not present with only metadata exchange filter, I have tried it.
It does not excercise the stats path.

I am not seeing these ASAN failures with both filters in the sidecars.
It happens consistently at ingress. Ingress has more traffic, more config and more config churn.

**Comment by mandarjog** [comment]
@jmarantz yes we can repro the use after free, but not in a unit or integration test.

**Comment by howardjohn** [comment]
@mandarjog just to clarify, I put heavy load on sidecars and see this as well -- so I think its just that the ingress has more load so its more obvious

**Comment by mandarjog** [comment]
The leak is there. We are talking about asan failure here.

**Comment by jmarantz** [comment]
@mandarjog are there detailed repro instructions then for a live server? Or maybe a fuzz test? Ideally it would be great to start with a failing test, otherwise how will we know if we fix it?

**Comment by mandarjog** [comment]
@jplevyak @kyessenov I used the latest istio proxy image. `gcr.io/istio-release/proxyv2:master-20191004-09-15` and did not see a memory leak.

**Comment by howardjohn** [comment]
Confirmed in master I don't see the leak

**Comment by howardjohn** [comment]
@mandarjog is the ASAN failure still occuring now?

**Comment by douglas-reid** [comment]
Can we close this issue now?

**Comment by mandarjog** [comment]
Still an issue in 1.3, but fixed in 1.4

Will keep open for discoverability. It is difficult to fix in 1.3 since code has diverged.

**Comment by howardjohn** [comment]
I don't think we should leave an issue open if it is fixed. you can still find closed issues
